NetCasbin/Rbac/DefaultRoleManager.cs

```diff
@@ -1,163 +1,271 @@
 ﻿using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 
 namespace NetCasbin.Rbac
 {
     public class DefaultRoleManager : IRoleManager
     {
-        private const string _domainError = "error: domain should be 1 parameter";
-        private readonly Dictionary<string, Role> _allRoles;
+        private readonly string _defaultDomain = "casbin::default";
+
+        private readonly ConcurrentDictionary<string, ConcurrentDictionary<string, Role>> _allDomains = new();
         private readonly int _maxHierarchyLevel;
 
+        public Func<string, string, bool> MatchingFunc { get; set; }
+        public Func<string, string, bool> DomainMatchingFunc { get; set; }
+
+        public bool HasPattern => MatchingFunc is not null;
+        public bool HasDomainPattern => DomainMatchingFunc is not null;
+
         public DefaultRoleManager(int maxHierarchyLevel)
         {
-            _allRoles = new Dictionary<string, Role>();
+            _allDomains.TryAdd(_defaultDomain, new ConcurrentDictionary<string, Role>());
             _maxHierarchyLevel = maxHierarchyLevel;
         }
 
-        private bool HasRole(string name)
+        #region Obsoleted APIs
+        public virtual List<string> GetRoles(string name, params string[] domain)
         {
-            return _allRoles.ContainsKey(name);
+            CheckDomainArgument(domain);
+            IEnumerable<string> roles = domain.Length is 0
+                ? GetRoles(name)
+                : GetRoles(name, domain[0]);
+            return roles as List<string> ?? roles.ToList();
         }
 
+        public virtual List<string> GetUsers(string name, params string[] domain)
```

Comment:
`virtual` here is not very meaningful, most people may not know how to override it correctly

Reply:
This is the design style now, we can remove it at the mainline version later.